lib/logstash/outputs/google_pubsub.rb

```diff
@@ -0,0 +1,145 @@
+#
+# === Environment Configuration
+#
+# To use this plugin, you must create a
+# https://developers.google.com/storage/docs/authentication#service_accounts[service account]
+# and grant it the publish permission on a topic.
+# The topic __must__ exist before you run the plugin.
+#
+# === Usage
+#
+# There are two major ways to authenticate to Google Cloud Pubsub:
+#
+# * autodiscovered
+#   https://cloud.google.com/docs/authentication/production[Application Default Credentials]
+# * a GCP Service Account JSON file
+#
+# Example configuration:
+#
+# [source,ruby]
+# ------------------------------------------------------------------------------
+# output {
+#   google_pubsub {
+#     # Required attributes
+#     project_id => "my_project"                                      (required)
+#     topic => "my_topic"                                             (required)
+#
+#     # Optional if you're using app default credentials
+#     json_key_file => "service_account_key.json"                     (optional)
+#
+#     # Options for configuring the upload
+#     message_count_threshold => 100                                  (optional)
+#     delay_threshold_secs => 5                                       (optional)
+#     request_byte_threshold => 4096                                  (optional)
+#     attributes => {"key" => "value"}                                (optional)
+#   }
+# }
+# ------------------------------------------------------------------------------
+#
+class LogStash::Outputs::GooglePubsub < LogStash::Outputs::Base
+  config_name 'google_pubsub'
+
+  concurrency :shared
+
+  # Google Cloud Project ID (name, not number)
+  config :project_id, validate: :string, required: true
+
+  # Google Cloud Pub/Sub Topic. You must create the topic manually.
+  config :topic, validate: :string, required: true
+
+  # If logstash is running within Google Compute Engine, the plugin will use
+  # GCE's Application Default Credentials. Outside of GCE, you will need to
+  # specify a Service Account JSON key file.
+  config :json_key_file, validate: :path, required: false
+
+  # Send the batch once this delay has passed, from the time the first message
+  # is queued. (> 0, default: 5)
+  config :delay_threshold_secs, validate: :number, required: true, default: 5
+
+  # Once this many messages are queued, send all the messages in a single call,
+  # even if the delay threshold hasn't elapsed yet. (< 1000, default: 100)
+  config :message_count_threshold, validate: :number, required: true, default: 100
+
+  # Once the number of bytes in the batched request reaches this threshold,
+  # send all of the messages in a single call, even if neither the delay or
+  # message count thresholds have been exceeded yet.
+  #
+  # This includes full message payload size, including any attributes set.
+  #
+  # Default: 1MB
+  config :request_byte_threshold, validate: :number, required: true, default: 1_000_000
```

Comment:
Consider using `validate: :bytes` - this enables users to specify values such as `1MB` or `256KiB` in their configurations rather than numeric values.

No need for `required: true`.